src/main/java/org/nuxeo/ecm/platform/pdf/service/PDFTransformationServiceImpl.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.pdfbox.pdmodel.graphics.xobject.PDXObjectImage;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.impl.blob.FileBlob;
+import org.nuxeo.ecm.platform.pdf.PDFUtils;
+import org.nuxeo.ecm.platform.pdf.service.watermark.WatermarkProperties;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+import javax.imageio.ImageIO;
+import java.awt.*;
+import java.awt.geom.AffineTransform;
+import java.awt.geom.Point2D;
+import java.awt.geom.Rectangle2D;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @since 8.10
+ */
+public class PDFTransformationServiceImpl extends DefaultComponent
+        implements PDFTransformationService {
+
+    protected static final Log log = LogFactory.getLog(PDFTransformationServiceImpl.class);
+
+    protected static final String MIME_TYPE = "application/pdf";
+
+    @Override
+    public WatermarkProperties getDefaultProperties() {
+        return new WatermarkProperties();
+    }
+
+    @Override
+    public Blob applyTextWatermark(Blob input, String text, WatermarkProperties properties) {
+
+        // Set up the graphic state to handle transparency
+        // Define a new extended graphic state
+        PDExtendedGraphicsState extendedGraphicsState = new PDExtendedGraphicsState();
+        // Set the transparency/opacity
+        extendedGraphicsState.setNonStrokingAlphaConstant((float) properties.getAlphaColor());
+
+        try (PDDocument pdfDoc = PDDocument.load(input.getStream())) {
+
+            PDFont font = PDType1Font.getStandardFont(properties.getFontFamily());
+            float watermarkWidth = (float) (font.getStringWidth(text) * properties.getFontSize()
+                                / 1000f);
+            int[] rgb = PDFUtils.hex255ToRGB(properties.getHex255Color());
+
+            for (Object o : pdfDoc.getDocumentCatalog().getAllPages()) {
+                PDPage page = (PDPage) o;
+                PDRectangle pageSize = page.findMediaBox();
+                PDResources resources = page.findResources();
+
+                // Get the defined graphic states.
+                HashMap<String, PDExtendedGraphicsState> graphicsStateDictionary =
+                        (HashMap<String, PDExtendedGraphicsState>) resources.getGraphicsStates();
+                if (graphicsStateDictionary != null) {
+                    graphicsStateDictionary.put("TransparentState",
+                            extendedGraphicsState);
+                    resources.setGraphicsStates(graphicsStateDictionary);
+                } else {
+                    Map<String, PDExtendedGraphicsState> m = new HashMap<>();
+                    m.put("TransparentState", extendedGraphicsState);
+                    resources.setGraphicsStates(m);
+                }
```

Comment:
This can probably be simplified by doing something like:

    Map<String, PDExtendedGraphicsState> graphicsStates = resources.getGraphicsStates();
    if (graphicsStates == null) {
        graphicsStates = new HashMap<>();
    }
    graphicsStates.put("TransparentState", extendedGraphicsState);
    resources.setGraphicsStates(graphicsStates);